src/details/ArborX_DetailsDistributor.hpp

```diff
@@ -31,6 +31,177 @@ namespace ArborX
+                              ++update.batch_count;
+                              update.total_count += batched_counts(i);
+                            }
+                          },
+                          result);
+    batch_offset += result.batch_count;
+    if (result.total_count > 0)
+    {
+      total_offset += result.total_count;
+      unique_ranks.push_back(largest_rank);
+      offsets.push_back(total_offset);
+    }
+  }
+  counts.reserve(offsets.size() - 1);
+  for (unsigned int i = 1; i < offsets.size(); ++i)
+    counts.push_back(offsets[i] - offsets[i - 1]);
+  ARBORX_ASSERT(unique_ranks.size() == counts.size());
+  ARBORX_ASSERT(offsets.size() == unique_ranks.size() + 1);
+
+  // step 2: prepare for creating the total permutation by reordering batch
+  //         counts and computing the new exclusive cumulative sum for these
+  Kokkos::View<int *, DeviceType> device_batched_permutation_indices_inverse(
+      Kokkos::ViewAllocateWithoutInitializing(
+          "batched_permutation_indices_inverse"),
+      batched_ranks.size());
+  Kokkos::parallel_for(
+      ARBORX_MARK_REGION("invert_batched_permutation"),
+      Kokkos::RangePolicy<ExecutionSpace>(0, batched_ranks.size()),
+      KOKKOS_LAMBDA(int i) {
+        device_batched_permutation_indices_inverse(
+            device_batched_permutation_indices(i)) = i;
+      });
+
+  InputView reordered_batched_counts(
+      Kokkos::ViewAllocateWithoutInitializing("reordered_batched_counts"),
+      batched_offsets.size()); // the last entry is unused
+  Kokkos::parallel_for(
+      ARBORX_MARK_REGION("reorder_batch_counts"),
+      Kokkos::RangePolicy<ExecutionSpace>(0, batched_offsets.size() - 1),
+      KOKKOS_LAMBDA(int j) {
+        reordered_batched_counts(j) =
+            batched_counts(device_batched_permutation_indices_inverse(j));
+      });
+
+  InputView exclusive_sum_reordered_batched_offsets(
+      Kokkos::ViewAllocateWithoutInitializing(
+          "exclusive_sum_reordered_batched_offsets"),
+      batched_offsets.size());
+  ArborX::exclusivePrefixSum(reordered_batched_counts,
+                             exclusive_sum_reordered_batched_offsets);
+
+  // step 3: compute the total permutation
+  Kokkos::View<int *, DeviceType> device_permutation_indices_inverse(
+      Kokkos::ViewAllocateWithoutInitializing(
+          "device_permutation_indices_inverse"),
+      permutation_indices.size());
+
+  Kokkos::parallel_for(
+      ARBORX_MARK_REGION("set_permutation_indices"),
+      Kokkos::RangePolicy<ExecutionSpace>(
+          0, device_batched_permutation_indices.size()),
+      KOKKOS_LAMBDA(int i) {
+        int n_batch_entries = reordered_batched_counts[i];
+        for (int j = 0; j < n_batch_entries; ++j)
+        {
+          device_permutation_indices_inverse(
+              exclusive_sum_reordered_batched_offsets(i) + j) =
+              batched_offsets(device_batched_permutation_indices_inverse(i)) +
+              j;
+        }
+      });
```

Comment:
Look into the function to apply permutations in `src/details/ArborX_DetailsBatchedQueries.hpp`.  I think there is opportunities for code reuse and improved readability.